tests/integrations/opentelemetry/test_sampler.py

```diff
@@ -0,0 +1,351 @@
+        ]
+
+        assert transaction_a["transaction"] == "request a"
+        assert transaction_b["transaction"] == "request b"
+
+        spans_a = transaction_a["spans"]
+        assert len(spans_a) == 2
+        assert spans_a[0]["description"] == "cache a"
+        assert spans_a[1]["description"] == "db a"
+        spans_b = transaction_b["spans"]
+        assert len(spans_b) == 2
+        assert spans_b[0]["description"] == "cache b"
+        assert spans_b[1]["description"] == "db b"
+
+
+def test_sampling_traces_sample_rate_50(init_sentry_with_potel, capture_envelopes):
+    init_sentry_with_potel(traces_sample_rate=0.5)
+
+    envelopes = capture_envelopes()
+
+    # Make sure random() always returns the same values
+    with mock.patch(
+        "sentry_sdk.integrations.opentelemetry.sampler.random",
+        side_effect=[0.7, 0.7, 0.7, 0.2, 0.2, 0.2],
+    ):
+        with sentry_sdk.start_span(description="request a"):
+            with sentry_sdk.start_span(description="cache a"):
+                with sentry_sdk.start_span(description="db a"):
+                    ...
+
+        with sentry_sdk.start_span(description="request b"):
+            with sentry_sdk.start_span(description="cache b"):
+                with sentry_sdk.start_span(description="db b"):
+                    ...
+
+    assert len(envelopes) == 1
+
+    (envelope,) = envelopes
+    transaction = envelope.items[0].payload.json
+    assert transaction["transaction"] == "request b"
+    spans = transaction["spans"]
+    assert len(spans) == 2
+    assert spans[0]["description"] == "cache b"
+    assert spans[1]["description"] == "db b"
+
+
+def test_sampling_traces_sampler(init_sentry_with_potel, capture_envelopes):
+    def keep_only_a(sampling_context):
+        if " a" in sampling_context["transaction_context"]["name"]:
+            return 0.05
+        else:
+            return 0
+
+    init_sentry_with_potel(
+        traces_sample_rate=1.0,
+        traces_sampler=keep_only_a,
+    )
+
+    envelopes = capture_envelopes()
+
+    # Make sure random() always returns the same values
+    with mock.patch(
+        "sentry_sdk.integrations.opentelemetry.sampler.random",
+        side_effect=[0.04 for _ in range(12)],
+    ):
+
+        with sentry_sdk.start_span(description="request a"):  # keep
+            with sentry_sdk.start_span(description="cache a"):  # keep
+                with sentry_sdk.start_span(description="db a"):  # keep
+                    ...
+
+        with sentry_sdk.start_span(description="request b"):  # drop
+            with sentry_sdk.start_span(description="cache b"):  # drop
+                with sentry_sdk.start_span(description="db b"):  # drop
+                    ...
+
+        with sentry_sdk.start_span(description="request c"):  # drop
+            with sentry_sdk.start_span(
+                description="cache a c"
+            ):  # keep (but trx dropped, so not collected)
```

Comment:
I don't like these mixed span drop/keep tests, the children span should just derive from the root span

Reply:
When you have a `traces_sampler` it is run for every span and you can ignore the parent sampling decision in your custom `traces_sampler`. This is what I wanted to show here.

Reply:
ah ok, all good then